### PR TITLE
Refactor Controller

### PR DESCRIPTION
**Pull request by haythem96**
# Refactor Controller

## High Level Description

This PR refactor the `redeem()`, `_settleVault()` and `isSettlementAllowed()` functions to reduce external calls.

Screen Shot 2021-01-11 at 1 17 15 AM

### Code

- [x] Unit test 100% coverage
- [x] Does your code follow the naming and code documentation guidelines?

### Documentation

- [ ] Is your code up to date with the spec? 
- [ ] Have you added your tests to the testing doc?

### CODE REVIEW

**Comment by haythem96** [comment]
Still need to figure out why Certora is failing.

File: contracts/Controller.sol [hunk 1]
```diff
@@ -398,21 +398,21 @@ contract Controller is Initializable, OwnableUpgradeSafe, ReentrancyGuardUpgrade
...
-
-        bool isUnderlyingFinalized = oracle.isDisputePeriodOver(underlying, expiry);
-        bool isStrikeFinalized = oracle.isDisputePeriodOver(strike, expiry);
-        bool isCollateralFinalized = oracle.isDisputePeriodOver(collateral, expiry);
+    function isSettlementAllowed(
+        address _underlying,
+        address _strike,
+        address _collateral,
+        uint256 _expiry
+    ) public view returns (bool) {
```
Comment (aparnakr) [inline contracts/Controller.sol:412]: hmm this may not be the most user-friendly. Should we user test this with devs first?

**Review by aparnakr — APPROVED** [review]
This looks good, the first refactor can we just user test with devs before merging?

File: contracts/Controller.sol (hunk 1 above)
Comment (haythem96) [reply contracts/Controller.sol:412]: Why it is not ?

**Review by antoncoding — APPROVED** [review]

**Pull request merged**